src/ctnetlink/nlas/flow/protocol_info.rs

```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: MIT
+
+use std::convert::TryFrom;
+
+use byteorder::{ByteOrder, NativeEndian};
+use netlink_packet_utils::{
+    nla::{Nla, NlaBuffer, NLA_F_NESTED},
+    DecodeError, Emitable, Parseable,
+};
+
+use crate::{
+    constants::{
+        CTA_PROTOINFO_DCCP, CTA_PROTOINFO_SCTP, CTA_PROTOINFO_TCP,
+        CTA_PROTOINFO_TCP_FLAGS_ORIGINAL, CTA_PROTOINFO_TCP_FLAGS_REPLY,
+        CTA_PROTOINFO_TCP_STATE, CTA_PROTOINFO_TCP_WSCALE_ORIGINAL,
+        CTA_PROTOINFO_TCP_WSCALE_REPLY, CTA_PROTOINFO_UNSPEC,
+    },
+    ctnetlink::nlas::ct_attr::{CtAttr, CtAttrBuilder},
+};
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum ProtocolInfo {
+    Tcp(ProtocolInfoTcp),
+    Dccp(CtAttr),
+    Sctp(CtAttr),
+    Other(CtAttr),
+}
+
+impl ProtocolInfo {
+    pub(super) fn parse_from_bytes(
+        buf: &[u8],
+    ) -> Result<ProtocolInfo, DecodeError> {
+        let b = NlaBuffer::new(buf);
+        ProtocolInfo::parse(&b)
+    }
+}
+
+impl Nla for ProtocolInfo {
+    fn value_len(&self) -> usize {
+        match self {
+            ProtocolInfo::Tcp(info) => info.buffer_len(),
+            ProtocolInfo::Dccp(attr) => attr.buffer_len(),
+            ProtocolInfo::Sctp(attr) => attr.buffer_len(),
+            ProtocolInfo::Other(attr) => attr.buffer_len(),
+        }
+    }
+
+    fn kind(&self) -> u16 {
+        match self {
+            ProtocolInfo::Tcp(_) => CTA_PROTOINFO_TCP | NLA_F_NESTED,
+            ProtocolInfo::Dccp(_) => CTA_PROTOINFO_DCCP | NLA_F_NESTED,
+            ProtocolInfo::Sctp(_) => CTA_PROTOINFO_SCTP | NLA_F_NESTED,
+            ProtocolInfo::Other(_) => CTA_PROTOINFO_UNSPEC,
+        }
+    }
+
+    fn emit_value(&self, buffer: &mut [u8]) {
+        match self {
+            ProtocolInfo::Tcp(info) => info.emit(buffer),
+            ProtocolInfo::Dccp(attr) => attr.emit(buffer),
+            ProtocolInfo::Sctp(attr) => attr.emit(buffer),
+            ProtocolInfo::Other(attr) => attr.emit(buffer),
+        }
+    }
+}
+
+impl<'buffer, T: AsRef<[u8]> + ?Sized> Parseable<NlaBuffer<&'buffer T>>
+    for ProtocolInfo
+{
+    fn parse(buf: &NlaBuffer<&'buffer T>) -> Result<Self, DecodeError> {
+        let attr = CtAttr::parse(buf)?;
+
+        match attr.attr_type {
+            CTA_PROTOINFO_TCP => {
+                Ok(ProtocolInfo::Tcp(ProtocolInfoTcp::try_from(attr)?))
+            }
+            CTA_PROTOINFO_DCCP => Ok(ProtocolInfo::Dccp(attr)),
+            CTA_PROTOINFO_SCTP => Ok(ProtocolInfo::Sctp(attr)),
+            _ => Ok(ProtocolInfo::Other(attr)),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default)]
+pub struct ProtocolInfoTcp {
+    pub state: u8,
+    pub wscale_original: u8,
+    pub wscale_reply: u8,
+    pub flgas_original: u16,
+    pub flags_reply: u16,
+}
```

Comment:
Please consider to use `NlaBuffer`, example: https://github.com/rust-netlink/netlink-packet-route/blob/main/src/route/via.rs#L31